HIP/hip-435.md

```diff
@@ -0,0 +1,131 @@
+Externalising verbose information about transaction executions without bloating the main TransactionRecord object with potentially unnecessary information and requiring mirror nodes to always download and process that information.
+
+## Rationale
+
+There are use-cases requiring a detailed and verbose output for a HAPI transaction execution. Users of Hedera should have the option to query a mirror node for this information. Meanwhile, bloating the TransactionRecords with information that would be unnecessary in most cases is not an ideal option. It is a much better design decision to create sidecar records that can be downloaded and processed only by interested parties.
+
+**Design Goal #1**
+Mirror nodes should not be required to download and process verbose transaction information. The sidecar transaction records will be externalised in separate record files.
+
+**Design Goal #2**
+One transaction can have 0, 1 or more than 1 sidecar records related to it.
+
+**Design Goal #3**
+It should be possible and relatively easy to add new types of sidecar transaction records.
+
+## User stories
+
+As a Hedera user, I want the to have the ability to access verbose information about transactions.
+
+As an operator of a mirror node, I want the ability to download and process only the most important record-related information.
+
+As an operator of a mirror node, I want the ability to download all the information about records execution that a Hedera node can provide.
+
+## Specification
+
+Introduce a sidecar transaction record protobuf message and a sidecar record file protobuf message to enable tracking of additional information about records execution.
+
+### TransactionSideCarRecord
+
+`TransactionSideCarRecord` is a protobuf which will be used to create sidecar records complementing `TransactionRecord` and storing additional information about a transaction execution.
+The `TransactionSideCarRecord`s created by Hedera nodes will be externalised in a new file, part of the new RecordStream structure described bellow.
+
+```
+message TransactionSideCarRecord {
+    // Consensus timestamp will be the same as the consensus timestamp of the 
+    // transaction the side car is related to. This offers a convenient 
+    // way to match record to sidecar. 
+    Timestamp consensusTimestamp = 1;
+    oneof {
+      repeated ContractStateChange = 2; // Example sidecar type
+      repeated ContractAction = 3; // Example sidecar type
+      // Other future categories
+    }
+}
+```
+
+### Record Stream File Changes
+
+A new version (V6) of the record stream file format will be created using the following protobufs:
+
+```
+message RecordFileObject {
+  TransactionRecord record = 1;
+  Transaction transaction = 2;
+}
+
+message RecordStreamFile {
+  bytes startObjectRunningHash = 1;
+  repeated RecordFileObject fileObjects = 2;
+  bytes endObjectRunningHash = 3;
```

Comment:
The running hash only covers the fileObjects and there is no cryptographic integrity check for the side car files?

Reply:
The record signature file that is created for each record stream file signs the hash of the entire corresponding stream file. The list of sidecar file hashes is included in the record stream file. This way mirror nodes or any interested party can download the record stream file and all sidecar files and verify that:
1. repeated bytes SideCarFileHash is correct; 
2. the signature file signed the correct hash of the entire record stream file;

Added this to the HIP as well.

Reply:
I didn't want to add the list of sidecar hashes to the RecordFileObject itself in order to avoid requiring the mirror nodes to download that information as well.